linode/manager
# feat: [M3-8336] - Improve support ticket form pre-population and field labels

## Description 📝
This is an enhancement from M3-7673, but it could apply more generally in the future.

Often, a user is asked to contact support (and linked to a support ticket form) via an API error. We lacked a way to pass any of the form data that the user submitted when they received that API error to the support ticket. This would be useful in the case where we want to reference something the user tried to submit, but could not. (In the case of the Linode Create account limit error, this is the linode plan type they selected). Having this data passed to the support ticket would allow us to prefill additional fields, which is less burden on the user.

## Changes  🔄
- Pre-populates the support ticket `linodePlan` field when an account limit ticket is opened with the user's plan selection from the form Linode and LKE create forms
- Updates test coverage to account for this change

And a few small refinements:
- *Customer/support facing*: handles the dynamic naming of `entities` in the field labels / some way to tell which type of entities are being requested
- *Customer/support facing*: fixes inconsistent ordering of fields as they appear in the ticket dialog vs. in the ticket body description of a submitted ticket (e.g. company was displaying before customer name)
- *Tech debt*: handles all prefilled data (via props or `Link` state parameters) in the SupportTicketDialog, not in SupportTicketLanding

## Target release date 🗓️
9/3

## Preview 📷
**Include a screenshot or screen recording of the change**

:bulb: Use `<video src="" />` tag when including recordings in table.

| Product | Before  | After   |
| --- | ------- | ------- |
| Linode | ![Screenshot 2024-08-14 at 1 13 54 PM](https://github.com/user-attachments/assets/4685dd03-cbde-473c-a841-008b0bf4e78e) ![Screenshot 2024-08-14 at 1 15 29 PM](https://github.com/user-attachments/assets/5ec648a3-cfea-47e2-a2ca-906041cd8c83) | <video src="https://github.com/user-attachments/assets/6d1cc7cc-45ff-4f15-9098-0f5d05829155"/> |
| LKE | basically same as Linodes | <video src="https://github.com/user-attachments/assets/18c0586b-bbfe-4490-88b5-4d844dec0fd7" /> |





## How to test 🧪

### Prerequisites
(How to setup test environment)
- **Please confirm this with the Linode Create Refactor flag ON and OFF.**
- **Please test this on your dev account so we don't bother Customer Support with real tickets.**
- Check out this PR and switch your environment to dev.
- Log into admin and set your thing limit to 0 on your dev account.

### Reproduction steps
(How to reproduce the issue, if applicable)
- To see the out of order fields, look at any submitted past tickets (SMTP or Account Limit), or create one in dev.
- To see the generic "entities" in the description, look at any past Account Limit ticket or create one in dev.

### Verification steps
(How to verify changes)
- Try to create a new Linode when exceeding your thing limit and confirm an Account Limits ticket pre-populates the plan field.
- Try to create a new Kube Cluster with multiple node pools when exceeding your thing limit and confirm an Account Limits ticket pre-populates the plan field with multiple plan types.
- Confirm the fields are in the same order in the ticket dialog as they are the submitted ticket description.
- Confirm that rather than `A detailed description of your use case and why you need access to more/larger entities` and `Total number of entities you need?`, the word `entities` is replaced with the correct entity type in both the support ticket form and the submitted ticket details.
- Confirm no regressions to regular support tickets and prefilled fields (for example, close out of, rather than cancel, a general support ticket and confirm data loads from local storage).
- Ensure tests pass:
```
yarn test src/features/Support/SupportTickets/ticketUtils.test.ts
```
```
yarn cy:run -s "cypress/e2e/core/helpAndSupport/open-support-ticket.spec.ts"
```

## As an Author I have considered 🤔

*Check all that apply*
